server/server/src/actions/exec.ts

```diff
@@ -0,0 +1,88 @@
+import { Action, Context } from '@anticrm/action'
+import type { ExecutionContext as CompleteExecutionContext } from '@anticrm/action-plugin'
+import type { Doc } from '@anticrm/core'
+import { getResource } from '@anticrm/platform'
+import { Resource } from '@anticrm/status'
+
+import { Service } from './service'
+
+type ExecutionContext = Omit<CompleteExecutionContext, keyof Doc>
+
+export async function exec (
+  { program, labelMap }: Action,
+  svc: Service,
+  ctx: ExecutionContext = { stack: [], counter: [0] },
+  parentCnt: number[] = []
+): Promise<void> {
+  const stack = [...ctx.stack]
+  let counter = ctx.counter[0]
+
+  while (counter < program.length) {
+    const atom = program[counter]
+    let recur = false
+
+    const atomCtx: Context = {
+      recur: async (ret) => {
+        stack.push(ret)
+        await svc.commit()
+        await svc.saveContext({ stack, counter: [...parentCnt, counter] })
+
+        recur = true
+      },
+      pop: () => stack.pop(),
+      push: (item: any) => stack.push(item),
+      create: svc.create.bind(svc),
+      update: svc.update.bind(svc),
+      subscribe: svc.subscribe.bind(svc),
+      findAll: svc.findAll.bind(svc)
+    }
+
+    if (atom._type === 'goto') {
```

Comment:
Could you have map with types and handlers?

Reply:
Let's update it a bit later, it doesn't seem to become more simple without changing rest of logic: handler context would be huge